### ISSUE TITLE
Allow "remap-lang" to run in immutable contexts

### ISSUE THREAD

**Issue by JeanMertz** [timberio/vector#4744]
The functionality in Remap is implemented through the `Expression` trait:

```rust
pub trait Expression: Send + Sync + std::fmt::Debug {
    fn execute(&self, state: &mut State, object: &mut dyn Object) -> Result<Option<Value>>;
}
```

Three implementation examples are (simplified code):

**assignment**

This is an expression implemented in remap-lang itself. It mutates the provided event (which implement `trait Object`) by assigning the given value to the path:

```rust
impl Expression for Assignment {
    fn execute(&self, state: &mut State, object: &mut dyn Object) -> Result<Option<Value>> {
        let value = self.value.execute(state, object)?;
		object.insert(&path, value.clone());

		Ok(Some(value))
    }
}
```

In the source code, this is written as:

```javascript
.foo = "bar"
```

**deletion**

This is a function that is not core to the language, but is provided by Vector itself:

```rust
impl Expression for DelFn {
    fn execute(&self, state: &mut State, object: &mut dyn Object) -> Result<Option<Value>> {
        for path in &self.paths {
            object.remove(path)
        }

        Ok(None)
    }
}
```

This mutates the event by removing a set of given paths.

```javascript
del(.foo)
```

**downcase**

The final example is an _immutable_ function, it manipulates a value, and returns the new value to be assigned, or returned as the final value of the source code:

```rust
impl Expression for DowncaseFn {
    fn execute(&self, state: &mut State, object: &mut dyn Object) -> Result<Option<Value>> {
        self.value
            .execute(state, object)?
            .map(|v| v.to_lowercase())
    }
}
```

In this example, the function itself doesn't mutate the event, but the assignment to `.foo` does:

```javascript
.foo = downcase(.foo)
```

## Problem

As can be seen in the above examples, all expressions require mutable access to the event, even if they don't mutate the event itself.

This provides problems in contexts where we don't want to allow event mutation. Specifically, in the context of conditionals (#4743), we don't have access to a mutable event, and even if we did, we explicitly want to disallow event mutation in this context.

## Solution

One possible solution to this problem is to split up the `Expression` trait into two traits:

```rust
pub trait Expression: Send + Sync + std::fmt::Debug {
    fn execute(&self, state: &mut State, object: &dyn Object) -> Result<Option<Value>>;
}

pub trait ExpressionMut: Send + Sync + std::fmt::Debug {
    fn execute(&self, state: &mut State, object: &mut dyn Object) -> Result<Option<Value>>;
}
```

Expressions that _don't_ mutate the event can implement both `Expression` and `ExpressionMut` (and delegate the second to the first). If an expression wants to mutate the event, it _can't_ implement `Expression`.

_(`State` can still be accessed mutably, since this contains internal state (e.g. variables), and doesn't relate to mutating the external world (e.g. events))_

You'd then be able to call `Program::new(source, functions)` or `Program::new_mut(source, functions)`. The former only accepts functions that implement `Expression`, the latter `ExpressionMut`.

For built-in expressions such as `Assignment`, we'd update the parser to reject the assignment expression if `new` is called instead of `new_mut`, which would produce a parse error for `.foo = "bar"`, signalling assignment is only allowed in a mutable-context.

**Comment by jszwedko** [comment]
:+1: I generally think this makes sense.

> For built-in expressions such as Assignment, we'd update the parser to reject the assignment expression if new is called instead of new_mut, which would produce a parse error for .foo = "bar", signalling assignment is only allowed in a mutable-context.

I agree that we shouldn't allow assignment of event fields, but would we want to still allow assignment of variables? I could see that being useful even in immutable contexts like the `reduce` transform.

**Comment by JeanMertz** [comment]
@jszwedko I agree about variables, which is why I (somewhat unclearly) mentioned it here:

> _(`State` can still be accessed mutably, since this contains internal state (e.g. variables), and doesn't relate to mutating the external world (e.g. events))_

So yes, I agree! Variables represent internal state, and will remain accessible (and assignable), even in an immutable context.

**Comment by jszwedko** [comment]
Ah! :smile: I missed that.

**Comment by binarylogic** [comment]
@JeanMertz is this still relevant given that recent changes?

**Comment by JeanMertz** [comment]
It is. Nothing has been done here. We have a work-around in place, but it would still be nice to get this in at some point (but there's no rush).
